src/ray/gcs/redis_context.cc

```diff
@@ -133,6 +135,71 @@ RedisContext::~RedisContext() {
   }
 }
 
+void GetRedisShards(redisContext *context, std::vector<std::string> *addresses,
+                    std::vector<int> *ports) {
+  // Get the total number of Redis shards in the system.
+  int num_attempts = 0;
+  redisReply *reply = NULL;
+  while (num_attempts < RayConfig::instance().redis_db_connect_retries()) {
+    // Try to read the number of Redis shards from the primary shard. If the
+    // entry is present, exit.
+    reply = reinterpret_cast<redisReply *>(redisCommand(context, "GET NumRedisShards"));
```

Comment:
Add a function std::unique_ptr< redisReply, std::function<void(redisReply*)>> ExecutionCommand(context, const char* command) {
    return std::unique_ptr< redisReply, std::function<void(redisReply*)>>(reinterpret_cast<redisReply *>(redisCommand(context, command)), [](redisReply *reply) {freeReplyObject(reply);});
}
to avoid manually call freeReplyObject(reply);